johnnydep/compat.py

```diff
@@ -1,17 +1,46 @@
 import oyaml
 
-
 try:
-    from urllib.request import urlretrieve
+    from urllib.parse import quote, urlparse
 except ImportError:
     # Python 2
-    from urllib import urlretrieve
+    from urllib2 import quote
+    from urlparse import urlparse
+
+
+def urlretrieve(url, filename, data=None, auth=None):
+    try:
+        from urllib import request as urllib2
+    except ImportError:
+        # Python 2
+        import urllib2
+
+    if auth is not None:
+        # https://docs.python.org/2.7/howto/urllib2.html#id6
+        password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
+
+        # Add the username and password.
+        # If we knew the realm, we could use it instead of None.
+        username, password = auth
+        top_level_url = urlparse(url).netloc
+        password_mgr.add_password(None, top_level_url, username, password)
+
+        handler = urllib2.HTTPBasicAuthHandler(password_mgr)
+
+        # create "opener" (OpenerDirector instance)
+        opener = urllib2.build_opener(handler)
+    else:
+        opener = urllib2.build_opener()
+
+    res = opener.open(url, data=data)
+
+    headers = res.info()
+
+    with open(filename, "wb") as fp:
+        fp.write(res.read())
+
+    return filename, headers
```

Comment:
so either py3 has urllib.request.urlretrieve, or py2 has urllib.urlretrieve (but no basicauth possibility), so now for py2 a solution using urllib2 (with basicauth) which imitates urllib.urlretrieve. working fine for private pypi on my side like this

it's of course possible extend the function to read/write in blocks (like in [official](https://github.com/wklken/Python-2.7.8/blob/c2336038b7f1ea84b8ae287e17381acb9ea438ff/Lib/urllib.py#L268) urlretrieve), but I'd say let's keep this function as uncluttered as possible

Reply:
OK, this is looking better now. Could you add test coverage for the branches which use auth in index url please?

Also remove the `.gitignore`, I prefer to one of those globally and only use per-project gitignore files for anything specifically generated by that project.

Reply:
I've added tests for different scenarios with/without auth in the index urls, and addressed the imports and gitignore. Good idea to move common stuff to a global gitignore!

Reply:
Cheers. I'll pull this branch down over the weekend and check it out, if all looks good will make a release for it.